src/Cli/Microsoft.DotNet.FileBasedPrograms/FileLevelDirectiveHelpers.cs

```diff
@@ -258,7 +258,30 @@ internal readonly record struct SourceFile(string Path, SourceText Text)
     public static SourceFile Load(string filePath)
     {
         using var stream = File.OpenRead(filePath);
-        return new SourceFile(filePath, SourceText.From(stream, Encoding.UTF8));
+        // Detect BOM to determine the appropriate encoding
+        Encoding encoding = DetectEncoding(stream);
+        stream.Position = 0; // Reset stream position after BOM detection
+        return new SourceFile(filePath, SourceText.From(stream, encoding));
```

Comment:
It looks like `SourceText` can already detect the encoding. If we pass `encoding: null` to `SourceText.From`, we wouldn't need to do any detection ourselves.